### ISSUE TITLE
Generated ligand poses are not in the protein pocket

### ISSUE THREAD

**Issue by hjuinj** [luost26/3D-Generative-SBDD#6]
I ran the `sample.py` script and from the `pool.finished` object I get a list of `data`. For each `data` I call `reconstruct_from_generated` to get out a rdkit molecule with a 3D conformation. 

However, it seems the conformation is not in the same coordinate space as the protein pocket:

I sampled idx 50 and looked at the `samples_all.pt` file, the corresponding protein pocket I loaded is thus: `FKB1A_HUMAN_2_108_0/1d7j_A_rec_1tco_fk5_lig_tt_docked_0_pocket10.pdb`

I see the ligand (left) is far away from the pocket (right)
<img width="559" alt="image" src="https://user-images.githubusercontent.com/5151132/153719030-8c8fe2d7-d063-4fde-99f1-3b9f7b5e7638.png">

 Am I doing something wrong?

**Comment by luost26** [comment]
Hi,

The PDB file of data#50 should be:
```
  protein_filename='IMA1_HUMAN_68_497_0/4u5s_A_rec_4u54_3c5_lig_tt_min_0_pocket10.pdb',
  ligand_filename='IMA1_HUMAN_68_497_0/4u5s_A_rec_4u54_3c5_lig_tt_min_0.sdf',
  id=93345,
```

The ordering of data in the testset is not the same as the ordering of `test_set.tsv`

**Comment by YuxiuLiu** [comment]
> 

Hi, did you get the 3D conformation by sample*.pt file? Could you please give the code to open them?

**Comment by luost26** [comment]
> > 
> 
> Hi, did you get the 3D conformation by sample*.pt file? Could you please give the code to open them?

They are stored as RDKit molecule objects in the `.pt` file. The `.pt` file can be loaded by `torch.load`

**Comment by YuxiuLiu** [comment]
> > Hi, did you get the 3D conformation by sample*.pt file? Could you please give the code to open them?
> 
> They are stored as RDKit molecule objects in the `.pt` file. The `.pt` file can be loaded by `torch.load`

Sorry, are they saved by smiles format?

**Comment by hjuinj** [comment]
for me, the sample.pt stored in the `finished` array a list of pytorch geometric data object which can be turned into rdkit molecule using the function `reconstruct_from_generated`

**Comment by hjuinj** [comment]
A different question: I am playing around with training the main network, and I see the training loss has not really changed after a few thousands of iterations. How many iterations did you train the model? This was not mentioned in the manuscript. It seems excessive to go to the max_iters of 1e7?

**Comment by YuxiuLiu** [comment]
> for me, the sample.pt stored in the `finished` array a list of pytorch geometric data object which can be turned into rdkit molecule using the function `reconstruct_from_generated`

Sorry to bother you again, where did you find the function 'reconstruct_from_generated'?

**Comment by hjuinj** [comment]
> > for me, the sample.pt stored in the `finished` array a list of pytorch geometric data object which can be turned into rdkit molecule using the function `reconstruct_from_generated`
> 
> Sorry to bother you again, where did you find the function 'reconstruct_from_generated'?

by function search in the code base: https://github.com/luost26/3D-Generative-SBDD/blob/61662090f1335da2449271aeec0c028d65911ac7/utils/reconstruct.py#L447

**Comment by YuxiuLiu** [comment]
> for me, the sample.pt stored in the `finished` array a list of pytorch geometric data object which can be turned into rdkit molecule using the function `reconstruct_from_generated`

Sorry to bother you again, I do not know how to use the reconstruct_from_generated function to convert the 'pt' files to sdf conformation, could you please show me your code?

**Comment by luost26** [comment]
> A different question: I am playing around with training the main network, and I see the training loss has not really changed after a few thousands of iterations. How many iterations did you train the model? This was not mentioned in the manuscript. It seems excessive to go to the max_iters of 1e7?

Hi, a few thousands of iterations are far from enough. 
The pretrained main model is taken from the ~1,500,000th iteration (takes about 2.5 days using a single 2080ti GPU), and the pretrained frontier model is taken from ~1,000,000th iteration.

The training loss and validation loss curves look like:
<img width="692" alt="Training loss" src="https://user-images.githubusercontent.com/45375360/154073097-10d0c9f4-8a8d-4cc3-b714-4f42cc5185cd.png">
<img width="355" alt="Validation loss" src="https://user-images.githubusercontent.com/45375360/154073460-dcec5e6f-5c91-42e9-8a22-990146ca9183.png">

**Comment by luost26** [comment]
> > for me, the sample.pt stored in the `finished` array a list of pytorch geometric data object which can be turned into rdkit molecule using the function `reconstruct_from_generated`
> 
> Sorry to bother you again, I do not know how to use the reconstruct_from_generated function to convert the 'pt' files to sdf conformation, could you please show me your code?

Hi, I have updated `sample.py`.

The generated molecules will be now saved as SDF files and SMILES files, beside the original `.pt` format.

<img width="339" alt="SDF" src="https://user-images.githubusercontent.com/45375360/154195140-594ec4e8-ddc9-4f0b-8606-10cbc5459613.png">
